From bf8915152296e05d1f20c99a342b2cc7f1e83a75 Mon Sep 17 00:00:00 2001
From: Al Stone <ahs3@redhat.com>
Date: Mon, 10 Nov 2014 17:11:20 -0700
Subject: [PATCH 099/131] clocksource: arm_arch_timer: fix system hang
Git-commit: d5c0f5afc6db4097ee36dd5ddaf212e958312728
Patch-mainline: Queued in subsystem maintainer repository
Git-repo: git://git.kernel.org/pub/scm/linux/kernel/git/rric/linux.git thunder/master-v4.1

Arm allows for two possible architectural clock sources. One memory mapped
and the other coprocessor based. If both timers exist, then the driver waits
for both to be probed before registering a clocksource.

Commit c387f07e6205 ("clocksource: arm_arch_timer: Discard unavailable timers
correctly") attempted to fix a hang occurring when one of the two possible
timers had a device node, but was disabled. In that case, the second probe
would never occur and the system would hang without a clocksource being
registered.

Unfortunately, incorrect logic in that commit made things worse such that
a hang would occur unless both timers had a device node and were enabled.
This patch fixes the logic so that we don't wait to probe a second timer
unless it exists and is enabled.

Signed-off-by: Mark Salter <msalter@redhat.com>

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

---
 drivers/clocksource/arm_arch_timer.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/clocksource/arm_arch_timer.c b/drivers/clocksource/arm_arch_timer.c
index 1a2f8ad..3f90bbc 100644
--- a/drivers/clocksource/arm_arch_timer.c
+++ b/drivers/clocksource/arm_arch_timer.c
@@ -672,10 +672,11 @@ arch_timer_needs_probing(int type, const struct of_device_id *matches)
 	bool needs_probing = false;
 
 	dn = of_find_matching_node(NULL, matches);
-	if (dn && of_device_is_available(dn) && !(arch_timers_present & type))
-		needs_probing = true;
-	of_node_put(dn);
-
+	if (dn) {
+		if (dn && of_device_is_available(dn) && !(arch_timers_present & type))
+			needs_probing = true;
+		of_node_put(dn);
+	}
 	return needs_probing;
 }
 
-- 
1.7.12.4

